Improve device-tree overlay support

sorki opened the pull request:
Now allows applying external overlays either in form of
`.dts` file, literal dts contents added to store or precompiled `.dtbo` file.

If overlays are defined, kernel device-trees are compiled with `-@`
so the `.dtb` files contain symbols which we can reference in our
overlays.

Since `fdtoverlay` doesn't respect `/ compatible` by itself
we query compatible strings of both `dtb` and `dtbo(verlay)`
and apply only if latter is substring of the former.

Also adds support for filtering `.dtb` files (as there are now nearly 1k
dtbs).

Mashup of
- @georgewhewell's https://gist.github.com/georgewhewell/b359404255f3bce6e54f980b0c092702
- and #67989

Introduces a breaking change for users of `hardware.deviceTree.base` which is now renamed to `hardware.deviceTree.kernelPackage` and requires actual kernel package for `dtc` recompilation to work.

<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

sorki committed:
nixos/device-tree: improve overlays support

Now allows applying external overlays either in form of
.dts file, literal dts context added to store or precompiled .dtbo.

If overlays are defined, kernel device-trees are compiled with '-@'
so the .dtb files contain symbols which we can reference in our
overlays.

Since `fdtoverlay` doesn't respect `/ compatible` by itself
we query compatible strings of both `dtb` and `dtbo(verlay)`
and apply only if latter is substring of the former.

Also adds support for filtering .dtb files (as there are now nearly 1k
dtbs).

Co-authored-by: georgewhewell <georgerw@gmail.com>
Co-authored-by: Kai Wohlfahrt <kai.wohlfahrt@gmail.com>

samueldr reviewed (dismissed):
> Introduces a breaking change for users of hardware.deviceTree.base which is now renamed to hardware.deviceTree.kernelPackage and requires actual kernel package for dtc recompilation to work.

We'll need the release notes to show the breaking change.

samueldr commented on nixos/modules/hardware/device-tree.nix:
```diff
@@ -43,14 +152,27 @@ in {
+    in lib.singleton {
+      assertion = lib.all (o: !invalidOverlay o) cfg.overlays;
+      message = ''
+        deviceTree overlay needs one of dtsFile, dtsText or dtboFile set";
```
```suggestion
        deviceTree overlay needs one of dtsFile, dtsText or dtboFile set.
```

samueldr commented:
I don't have the setup to test, exactly, but I do have some hardware I should setup to test.

Otherwise, at a glance this looks fine, though not super thrilled about any breaking change, but the change is for the best considering it allows multiple representations of device trees to be used here.

sorki commented:
You can test on rPi with https://gist.github.com/sorki/c72dfbf1af51f7a42693622fbd020f07

I think that the previous `hardware.deviceTree.base` package was used mainly with `deviceTree_rpi` (which should probably be removed as well as in #67989). Provided you don't need to apply overlays you can still set `hardware.deviceTree.kernelPackage` to custom derivation. I guess that not many people are using their custom `dtbs` instead of one provided by kernel + overlays as that would require a lot of maintenance.

sorki commented:
> We'll need the release notes to show the breaking change.

Release note added

samueldr commented:
Ah, cc @bennofs who played with out of tree DTBs recently. Might have some thoughts about this.

samueldr commented on nixos/modules/hardware/device-tree.nix:
```diff
@@ -16,25 +110,40 @@ in {
           '';
         };
 
-        base = mkOption {
```
Oops, this went past me, since this is a removed option, well, one could argue renamed, we want to `mkRemovedOptionModule` with a really quick text pointing towards `hardware.deviceTree.kernelPackage` and to look at the 20.03 release notes.

The end-users would otherwise just see that the option doesn't exist, now they would see an explanation as to why and be orientated about what to do.

bennofs replied:
Do we even need this option? In what case would it make sense to use something different from `config.boot.kernelPackages.kernel` here?

kwohlfahrt replied:
I have used files from `raspberrypifw` in the past, though that caused a few issues with a mismatched kernel so I wouldn't necessarily recommend it.

sorki replied:
Keeping it there for now, it can be useful for 3rd party repositories containing overlays.

samueldr commented:
When I said "hardware to test" I meant I have a raspberry pi touch screen I can test this with, still hadn't looked at how to make it work. My main issue is that I would prefer the mainline kernel, but there's a dire lack of documentation about "raspberry pi special hardware" and mainline.

bennofs commented on nixos/modules/hardware/device-tree.nix:
```diff
@@ -4,6 +4,100 @@ with lib;
+        installPhase = ''
+          make dtbs_install INSTALL_DTBS_PATH=$out/dtbs
+        '';
+      };
```
Would it make sense to move this function, and maybe also `filterDTBs` and `compileDTS` to nixpkgs (next to `applyOverlays`)? I think these functions might sometimes be useful for users to build their own device tree packages, and having them inside the module makes it hard to reuse.

sorki replied:
`compileDTS` sounds good, not sure about `filterDTBs` as that is a pretty generic store path filter which resembles `filterSource` but for store paths..

sorki replied:
For now I'm keeping this internal, if you would like to have this exposed from `pkgs` let me know and I'll try to fit it somewhere but I'm bad at namespaces and naming :) It also needs a better name in that case, something like `pkgs.compileDeviceTreeOverlays`.

bennofs reviewed (approved):
I like it, especially the recompiling the kernel dts with symbols part :+1: Because I didn't have the patience to figure out how that works, I just compiled a completely new device tree for my own use case. 

I added two review comments, but I would also be fine with the current state of the PR.

kwohlfahrt reviewed (commented):
I won't be able to test this until Monday, but at first glance I like the `overlayType` and `filterDtbs`.

I don't like the testing for `/compatible`, I feel like trying to apply an overlay to an incompatible device-tree should give a visible error rather than being silently ignored. Any thoughts from others?

sorki commented:
> I don't like the testing for `/compatible`, I feel like trying to apply an overlay to an incompatible device-tree should give a visible error rather than being silently ignored. Any thoughts from others?

At first I've tried applying overlays to all devices but that results in tons of errors on the output as most are incompatible (I've thought `fdtoverlay` respects compatible automatically but that's not the case).

Then my second version had `overlay.match` POSIX regex for `find` but that looked needlessly complicated and redundant to `compatible` strings in `dts` itself so I've stripped that and used `fdtget` instead to query both. It now results in following log output:

```
Applying overlay spi to bcm2835-rpi-zero-w.dtb
Applying overlay pps to bcm2835-rpi-zero-w.dtb
Applying overlay spi to bcm2836-rpi-2-b.dtb
Applying overlay pps to bcm2836-rpi-2-b.dtb
...
```

Which I think is better than seeing bunch of errors due to incompatibilities.

kwohlfahrt commented:
Would it make sense to do the filtering in `filterDtbs`? Something like:

```
filterDtbs = {
  name = "...";
  compatible = "...";
}
```

And then only copying based on `fdtget` output?


Pros:
- Lets you avoid errors on applying overlays to device-trees you don't mean to
- Gives explicit errors when overlays unexpectedly don't apply due to compatibility

Cons:
- Need to specify compatibility twice (once in overlay, once in nixpkgs)
- Doesn't work if you have multiple overlays that are meant to apply to distinct subsets of your .dtb collection. I'm not sure if this is a common use-case?

I'm not 100% opposed to the current solution, it is very flexible.

sorki commented:
> Would it make sense to do the filtering in `filterDtbs`? Something like:

Not sure if that would be an improvement - often you want to patch only few files but preserve the rest, `filterDTBs` is good for saving some CPU time or if you only need `DTB`s for one specific board and don't care about the rest. I'm not a big fan of duplicating `compatible` + another filter which is why I've scraped that attempt - the implementation was also becoming too complicated compared to current state.

sorki commented:
Removed option is now addressed via `mkRemovedOptionModule`.

kwohlfahrt commented:
OK, I'm convinced by the current approach for `compatible`. I'll try and test it tomorrow.

georgewhewell reviewed (approved):
LGTM

samueldr commented:
Hmm, this is a bit annoying...

Either:

 * We have the authorization by @NixOS/nixos-release-managers to backport this feature change to 20.03.
 * The release notes are moved to 20.09.

Otherwise looks fine to me.

sorki commented:
I've moved release note to 20.09 and fixed one bug - if no filters or overlays were applied the default path was missing `/dtbs` suffix.

sorki commented on nixos/modules/hardware/device-tree.nix:
```diff
@@ -4,7 +4,106 @@ with lib;
+      };
+
+  filterDTBs = src: if isNull cfg.filter
+    then "${src}/dtbs"
```
Previously only `then src`, missing `/dtbs`

worldofpeace commented:
> Hmm, this is a bit annoying...
> 
> Either:
> 
>     * We have the authorization by @NixOS/nixos-release-managers to backport this mildly-breaking change to 20.03.
> 
>     * The release notes are moved to 20.09.
> 
> 
> Otherwise looks fine to me.

20.03 is starting to look like a lot of work to get to the official release :rofl: So I'd rather be super strict about freeze exceptions.

samueldr commented:
Anyway it looks like @sorki is fine with bumping to 20.09, thanks anyway @worldofpeace  :).

kwohlfahrt commented on pkgs/os-specific/linux/device-tree/default.nix:
```diff
@@ -1,16 +1,30 @@
+      mkdir -p $out
+      for dtb in $(find ${base} -name '*.dtb'); do
+        cp $dtb $out
+      done
```
It seems to me that uboot requires the same directory structure (e.g. `broadcom/*.dtb` instead of just `*.dtb`) as the kernel source. This code just copies the files into one flat directory. I propose a similar change as in `filterDTBs`:

```suggestion
      cd ${base}
      find . -type f -name '*.dtb' -print0 \
        | xargs -0 cp -v --no-preserve=mode --target-directory $out --parents
```

sorki replied:
While not strictly needed it's a nice improvement, switched to your suggestion.

kwohlfahrt reviewed (changes requested):
Just got around to testing this, my apologies for the huge delay. The API is good I think, but as-is this does not work for me (at least rebased onto 19.09). The changes suggested in the review fix it, let me know what you think?

Wraul commented on nixos/modules/hardware/device-tree.nix:
```diff
@@ -4,7 +4,106 @@ with lib;
+
+        installPhase = ''
+          make dtbs_install INSTALL_DTBS_PATH=$out/dtbs
+        '';
```
For cross-compiling I had to make the following changes.

- The `depsBuildBuild` were necessary or else I got an error stating that `bin/sh: gcc: not found`.
- The `ARCH` defines where necessary or else the `dtbs` target weren't available.

```suggestion
        inherit (cfg.kernelPackage) src nativeBuildInputs depsBuildBuild;
        patches = map (patch: patch.patch) cfg.kernelPackage.kernelPatches;

        buildPhase = ''
          substituteInPlace scripts/Makefile.lib \
            --replace 'DTC_FLAGS += $(DTC_FLAGS_$(basetarget))' 'DTC_FLAGS += $(DTC_FLAGS_$(basetarget)) -@'
          make ${pkgs.stdenv.hostPlatform.platform.kernelBaseConfig} ARCH="${pkgs.stdenv.hostPlatform.platform.kernelArch}"
          make dtbs  ARCH="${pkgs.stdenv.hostPlatform.platform.kernelArch}"
        '';

        installPhase = ''
          make dtbs_install INSTALL_DTBS_PATH=$out/dtbs  ARCH="${pkgs.stdenv.hostPlatform.platform.kernelArch}"
        '';
```

kwohlfahrt replied:
These changes look sensible to me, they are similar to what is done in the kernel package.

sorki replied:
Thanks!

Wraul commented on nixos/modules/hardware/device-tree.nix:
```diff
@@ -4,7 +4,106 @@ with lib;
+
+  compileDTS = name: f: pkgs.runCommand "${name}-dtbo" {} ''
+    ${pkgs.dtc}/bin/dtc -I dts ${f} -O dtb -@ -o $out
+  '';
```
For cross-compiling I had to make this change or else the ARM version  of `dtc` were used.

```suggestion
  compileDTS = name: f: pkgs.callPackage({dtc}: pkgs.runCommand "${name}-dtbo" {
    nativeBuildInputs = [dtc];
  } ''
    dtc -I dts ${f} -O dtb -@ -o $out
  '') {};
```

kwohlfahrt replied:
The principle here is OK too, I think if we're passing `nativeBuildInputs` we may as well use `mkDerivation` here with `buildCommand`.

sorki replied:
Fixed using `mkDerivation`.

Wraul reviewed (commented):
I'm trying to utilize this PR for my Raspberry pi 3. The way I'm doing things currently is by cross-compiling on a Ubuntu machine and deploying the result to the rpi using nixops.

The cross-compilation didn't work without a couple of changes, though. So I thought I should make you aware of this fact.

I have left two code comments which should show the changes I made.
Though I'm very new to Nix so this is most likely not the way to go about fixing this, but it did at least allow me to get through the compilation.

sorki commented on nixos/modules/hardware/device-tree.nix:
```diff
@@ -4,7 +4,106 @@ with lib;
+    then "${src}/dtbs"
+    else
+      pkgs.runCommand "dtbs-filtered" {} ''
+        mkdir -p $out
+        for f in $( find ${src} -maxdepth 2 -type f -name '${cfg.filter}' ); do
+          cp -v $f $out/
+        done
```
Integrating this with small change of `cd ${src}/dtbs" -> `cd ${src}` not to restrict `src` to have to contain `dtbs` subfolder.

sorki commented:
Thank you all for the review, I've addressed your suggestions and tested natively, cross build is still running.

If you're looking for mainline compatible overlays similar to `rasberrypi-fw` take a look at https://github.com/afaerber/dt-overlays 

While updating this I've realized it would be nice to have rPi specific modules and device tree overlays collected somewhere in nixpkgs but I'm not sure what would be a good namespace for those.

sorki commented:
Fixed an issue I've found today with `linux 5.7` by adding `patchShebangs` to `dtbsWithSymbols`.

doronbehar commented:
This PR looks really good and I have a raspberry Pi which now runs Debian because support for overlays was missing at the time I tried NixOS. I'm subscribed to updates here but I don't have time to test it yet. Never the less thanks for putting work into this @sorki :rocket:.

kwohlfahrt commented:
I've tested the new updates, still all good.

> If you're looking for mainline compatible overlays similar to rasberrypi-fw take a look at https://github.com/afaerber/dt-overlays

Good tip! I've been looking for exactly that.

kwohlfahrt commented:
Actually, I'm seeing serious issues with a missing fdt. I'm trying to track it down now, it might be due to me re-basing these patches wrong.

kwohlfahrt commented on nixos/modules/hardware/device-tree.nix:
```diff
@@ -4,7 +4,114 @@ with lib;
+    else
+      pkgs.runCommand "dtbs-filtered" {} ''
+        mkdir -p $out
+        cd ${src}
```
This also needs to be `${src}/dtbs`. Otherwise, filtered/overlaid dtbs end up in `/boot/nixos/$out/dtbs/vendor/chip.dtb`, which is not found by uboot (it searches `/boot/nixos/$out/vendor/chip.dtb`).

```suggestion
        cd ${src}/dtbs
```

kwohlfahrt replied:
You mentioned in a previous comment, that you don't want to require a `dtbs` subdirectory.

Alternatively, the input to `filterDTBs`/`applyOverlays` could be changed to something like `"${cfg.kernelPackage}/dtbs"` and `"${dtbsWithSymbols}/dtbs"`, but personally I think adding the subdirectory here is cleaner.

sorki replied:
Thanks, good catch! Juggling many things I sometimes forget what I did before :| Addressed.

kwohlfahrt commented on pkgs/os-specific/linux/device-tree/default.nix:
```diff
@@ -1,16 +1,30 @@
+        if [[ "$dtbCompat" =~ "$overlayCompat" ]]; then
+          echo "Applying overlay ${o.name} to $( basename $dtb )"
+          mv $dtb{,.in}
+          fdtoverlay -o "$dtb" -i "$dtb.in" ${o.dtboFile};
```
This leaves `$dtb.dtb.in` in the directory, I suggest cleaning it up with `rm $dtb.in` unless there's a reason for keeping it around?

sorki replied:
Done

kwohlfahrt reviewed (changes requested):
Unforunately, filtered and overlays are broken for me, the suggested patch fixes it.

Also, the original DTB files are left as `.dtb.in`. uboot cannot use these to boot from, so I suggest cleaning them up.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/raspberry-pi-4-add-i2c-to-device-tree/7380/2

sorki commented:
Rebased and added suggestions. Tested lightly on my armv7 notebook, will test with rPI later as my room is too hot right now for cross compiling :)

I've tried to make a NixOS test for this but it seems a bit contrived - build a VM with overlay applied and just boot it (and only applicable to `armv7l/aarch64`). Maybe we can expose few examples for Hydra to build instead but that also requires some restructuring I think.

kwohlfahrt commented:
I've been testing this (rebased onto 20.03 but otherwise unchanged). LGTM.

sorki commented:
Retested, found one issue where `patchShebangs` was missing an argument, now corrected to `patchShebangs scripts/*`.

Should be good to go now.

Infinisil commented on nixos/modules/hardware/device-tree.nix:
```diff
@@ -38,14 +145,29 @@ in {
           '';
         };
 
+        filter = mkOption {
+          type = types.nullOr types.str;
+          default = null;
+          example = "*rpi*.dtb";
+          description = ''
+            Only include .dtb files matching glob expression.
+          '';
+        };
+
         overlays = mkOption {
           default = [];
-          example = literalExample
-            "[\"\${pkgs.device-tree_rpi.overlays}/w1-gpio.dtbo\"]";
-          type = types.listOf types.path;
+          example = literalExample ''
+            [
+              { name = "pps"; dtsFile = ./dts/pps.dts; }
+              { name = "spi";
+                dtsText = "...";
+              }
+              { name = "precompiled"; dtboFile = ./dtbos/example.dtbo; }
+            ]
+          '';
+          type = types.listOf overlayType;
```
This is a breaking change like this. You can change the type to (untested)
```nix
types.listOf (types.coercedTo types.path (path: {
  name = baseNameOf path;
  dtboFile = path;
}) overlayType
```

for it to be backwards compatible

lopsided98 replied:
I think this would be better as an `attrsOf` because it is much easier to override, although that is even more of a breaking change.

lopsided98 replied:
Although I guess that makes establishing the order of the overlays more difficult.

Infinisil replied:
Hm true. I plan to make an RFC to deprecate `listOf` types (as good as possible) in favor of `attrsOf`. Ordering of values is one of the few things I haven't quite figured out yet.

sorki replied:
Looks good, thanks! I'll try to incorporate this and switch to `attrsOf` since the `name` is exactly `attr` name. Best to do it now rather than introducing another breaking change later.

kwohlfahrt replied:
I think the order in which overlays are applied could matter - for example if I include `vendor.dtb`, which adds a device with some sensible defaults, and then `customization.dtb`, which overwrites some of these with my specific configuration.

Though I have to admit I'm not actually doing this right now, but it's a use-case I could imagine. IMO it would be better to
leave this as a list until the ordering issue mentioned above is figured out.

kwohlfahrt replied:
I think the `name` attribute here is somewhat misleading, it doesn't specify anything about where the overlay is applied, or in what order. It's just an internal name given to the generated store path, so `attrsOf` is not the right type here.

sorki replied:
Good point, let's keep it as `listOf` for now with the added compatibility using `coercedTo`.

WRMilling commented:
@Infinisil worked with me yesterday to get this PR patched onto my `nixpkgs` install (I am running NixOS 20.09 on a Pinebook Pro). I used it to patch in a single `dtbo` file and it worked as expected applying only to the specific context I cared about.

kwohlfahrt commented:
I think the order in which overlays are applied could matter - for example
if I include `vendor.dtb`, which adds a device with some sensible defaults,
and then `customization.dtb`, which overwrites some of these with my
specific configuration. Though I have to admit I'm not actually doing this
right now, but it's a use-case I could imagine. IMO it would be better to
leave this as a list until the ordering issue mentioned above is figured
out.

On Fri, 7 Aug 2020 at 13:59, Richard Marko <notifications@github.com> wrote:

> *@sorki* commented on this pull request.
> ------------------------------
>
> In nixos/modules/hardware/device-tree.nix
> <https://github.com/NixOS/nixpkgs/pull/79370#discussion_r467024303>:
>
> > -          example = literalExample
> -            "[\"\${pkgs.device-tree_rpi.overlays}/w1-gpio.dtbo\"]";
> -          type = types.listOf types.path;
> +          example = literalExample ''
> +            [
> +              { name = "pps"; dtsFile = ./dts/pps.dts; }
> +              { name = "spi";
> +                dtsText = "...";
> +              }
> +              { name = "precompiled"; dtboFile = ./dtbos/example.dtbo; }
> +            ]
> +          '';
> +          type = types.listOf overlayType;
>
> Looks good, thanks! I'll try to incorporate this and switch to attrsOf
> since the name is exactly attr name. Best to do it now rather than
> introducing another breaking change later.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/79370#discussion_r467024303>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AASPNNRIU7X4A6Q4TN6ZFI3R7P3D5ANCNFSM4KRBKD5Q>
> .
>

lopsided98 commented:
I've been using this with a precompiled RPi firmware overlay as well as a custom one compiled from source and it has worked well.

doronbehar commented:
Thanks @Infinisil for being prompt to help. TBH I haven't ran a `nix-build` command yet since I had some trouble understanding the rewrite of the `overlays`  option. Now I do understand it (mostly), but I figured out that in order to build such an SD image with my x86_64 machine, I would need to setup a qemu builder and I haven't got there yet. This, plus the fact that I just bought a new Laptop, and I'd like to use it + my previous computer as nix builders and there's lots of other changes to these machines' configurations, has delayed my progress with this task. But I sure hope to reach there on the next few days and I'll report here if I experience any issues. Thanks again. 

1 detail that's still missing for me from this PR, is: How to add parameters to an overlay such as explained here:

https://github.com/raspberrypi/firmware/blob/031e8d26610874cc43a5726cb07feef15c919052/boot/overlays/README#L440-L458

sorki commented:
@doronbehar `qemu-user` / `binfmt` is not necessarily needed (and people report hangs of the VMs when building certain packages) 

You can try cross-compiling the image instead, using something like https://gist.github.com/sorki/548de08f621b066c94f0c36a7a78cc41 - unless you need graphics it's a viable way to build images, provided you pick working commit (master is almost working for me except for failing `texinfo` build due to `smartmontools` -> `mailutils` -> `texinfo` dep chain, where `smartmontools` is pulled from `nixos/modules/profiles/base.nix` which can be patched directly in checkout or overridden somehow properly I guess.

Regarding parameters I'm not quite sure how these are implemented but I've tried to decompile the one you've posted (https://github.com/raspberrypi/firmware/blob/master/boot/overlays/allo-piano-dac-plus-pcm512x-audio.dtbo) with `dtc -I dtb -O dts < allo-piano-dac-plus-pcm512x-audio.dtbo` and there's basically this part 

```
__overrides__ {
	24db_digital_gain = <0x00>;
	glb_mclk = <0x00>;
}
```

which could be templated using Nix if you provide the overlay as Nix string.

sorki commented:
Thank you all for testing. Updated with the suggestion by @Infinisil

kwohlfahrt commented:
IIRC device-tree parameters are a feature of the raspberry pi bootloader, and don't work with the standard device-tree toolchain (including `dtc`). I usually just copy the source for the overlay and manually insert the parameter.

I'd quite like to have this in for 20.09, is that looking likely?

doronbehar commented:
@kwohlfahrt:

> device-tree parameters are a feature of the raspberry pi bootloader, and don't work with the standard device-tree toolchain (including dtc). I usually just copy the source for the overlay and manually insert the parameter.

@sorki:

> basically this part [..] which could be templated using Nix if you provide the overlay as Nix string.

I don't know a thing about decompiling dt overlays and manually inserting such parameters. Do you think it'd be possible to add another option to the `overlays` submodule that will do what you described? You have contributed a lot already so I hope this is not too much to ask.

sorki commented:
> I don't know a thing about decompiling dt overlays and manually inserting such parameters. Do you think it'd be possible to add another option to the `overlays` submodule that will do what you described? You have contributed a lot already so I hope this is not too much to ask.

I'm not sure if it's even viable way to do it since `dtbo` files are products built for specific kernel version (and the ones that are part of `raspberrypi/linux` aren't compatible with mainline). 

I think it would be better to find the `dts` (device tree overlay source file) used to produce the `dtbo` and just template that. We might want to maintain such (upstream compatible) overlays (for example based on https://github.com/afaerber/dt-overlays) as part of `nixos-hardware`.

Preferably I would keep that as a future feature (when we figure it out) as I too would like this to land for 20.09.

doronbehar commented:
> I think it would be better to find the dts (device tree overlay source file) used to produce the dtbo and just template that.

I see. The overlay I am interested in is not open source so I don't think there is a dts... :/.

sorki commented:
> > I think it would be better to find the dts (device tree overlay source file) used to produce the dtbo and just template that.
> 
> I see. The overlay I am interested in is not open source so I don't think there is a dts... :/.

Found it - https://github.com/raspberrypi/linux/blob/rpi-5.4.y/arch/arm/boot/dts/overlays/allo-piano-dac-plus-pcm512x-audio-overlay.dts

Was looking for it yesterday but only found the `dtbo`.

infinisil reviewed (approved):
I can't speak for all the builders in there, but the module options look good to me! Unless somebody else has any concerns with this, I think this is good to merge

sorki commented:
@jonringer do you think we can get this merged and cherry-picked to 20.09?

jonringer commented:
ZHF hasn't really started yet, I think it would be okay if it got in soon (today). The release notes may need to changed slightly if there's conflicts

worldofpeace commented on nixos/modules/hardware/device-tree.nix:
```diff
@@ -4,7 +4,114 @@ with lib;
+          Literal DTS contents, overlay is applied to
+          each .dtb file matching "compatible" of the overlay.
+        '';
+        example = literalExample ''
+          /dts-v1/;
+          /plugin/;
+          / {
+                  compatible = "raspberrypi";
+                  fragment@0 {
+                          target-path = "/soc";
+                          __overlay__ {
+                                  pps {
+                                          compatible = "pps-gpio";
+                                          status = "okay";
+                                  };
+                          };
+                  };
+          };
+        '';
```
is this nix code?  asking because it indented way further than the standard 2 spaces

sorki replied:
No it isn't - it's a sample of a (dts) Device Tree Source file.

worldofpeace commented on nixos/modules/hardware/device-tree.nix:
```diff
@@ -4,7 +4,114 @@ with lib;
+      substituteInPlace scripts/Makefile.lib \
+        --replace 'DTC_FLAGS += $(DTC_FLAGS_$(basetarget))' 'DTC_FLAGS += $(DTC_FLAGS_$(basetarget)) -@'
+      make ${pkgs.stdenv.hostPlatform.platform.kernelBaseConfig} ARCH="${pkgs.stdenv.hostPlatform.platform.kernelArch}"
+      make dtbs  ARCH="${pkgs.stdenv.hostPlatform.platform.kernelArch}"
```
```suggestion
      make dtbs ARCH="${pkgs.stdenv.hostPlatform.platform.kernelArch}"
```

worldofpeace commented on nixos/modules/hardware/device-tree.nix:
```diff
@@ -4,7 +4,114 @@ with lib;
+  };
+
+  # this requires kernel package
+  dtbsWithSymbols = pkgs.stdenv.mkDerivation {
```
why can't this be in pkgs but we override `kernalPackage` in this module?

sorki replied:
It's considered internal (same as `compileDTS` and `filterDTBs`) as it's only useful for this module.

worldofpeace commented:
I'm cool with this being brought into alpha since we haven't reached the beta freeze just yet (and it was supposed to be in the last release anyways)

worldofpeace commented:
Be sure to direct requests to releases managers via the team @NixOS/nixos-release-managers

worldofpeace commented:
Ah wait, I think this commit should be 
`nixos/device-tree: improve support`

sorki commented:
> Ah wait, I think this commit should be
> `nixos/device-tree: improve support`

Changed to `nixos/device-tree: improve overlays support`, thanks!

worldofpeace commented:
@sorki Okay, merged.

timclassic commented:
I believe the merge of this PR has broken the "With GPU" instructions on the wiki (https://nixos.wiki/wiki/NixOS_on_ARM/Raspberry_Pi_4#With_GPU).  I get the following error now:

```
# nixos-rebuild switch
building Nix...
building the system configuration...
error:
Failed assertions:
- The option definition `hardware.deviceTree.base' in `/etc/nixos/configuration.nix' no longer has any effect; please remove it.
Use hardware.deviceTree.kernelPackage instead

(use '--show-trace' to show detailed location information)
```

I tried deleting the `hardware.deviceTree.base` option but now X complains that it can't find `/dev/dri/card0` and fails to start.  What should I should provide for the `hardware.deviceTree.kernelPackage` option to produce the same behavior provided by `base = pkgs.device-tree_rpi;`?

jonringer commented:
cc @samueldr @NixOS/exotic-platform-maintainers

lovesegfault commented:
I too am struggling with these changes. It's not clear to me how the wiki instructions for using the RPi4 GPU should be now, and separately I'm trying to apply a `dtbo` file but finding it hard. 
```
  hardware.deviceTree.overlays = [{
    name = "hyperpixel4";
    dtboFile = "${pkgs.hyperpixel4}/share/overlays/hyperpixel4.dtbo";
  }];
```
I thought this would be sufficient (with `hardware.deviceTree.enable = true`) but the `dtbo` is _not_ applied.

sorki commented:
During build, you should see a line saying `Applying overlay xyz to ...`. If that's not the case decompile the `.dtbo` file with `dtc -I dtb -O dts < some.dtbo` (using `dtc` from `nix-shell -p dtc`) and check its toplevel `compatible` directive which should say something like `compatible = "brcm,bcm2835";` or `compatible = "raspberrypi"`.

`vc4-fkms-v3d.dtbo` looks good to me, not sure about the hyperpixel one.

Setting a kernel package shouldn't be needed but I have some trouble reproducing this (wiki vp4 instructions) with `linuxPackages_rpi4` as it gives me an error during `make ... dtbs` cross compilation (kernel package itself seems to cross fine):
```
gcc: error: unrecognized argument in option '-mabi=aapcs-linux'
gcc: note: valid arguments to '-mabi=' are: ms sysv
gcc: error: unrecognized command line option '-mlittle-endian'
gcc: error: unrecognized command line option '-mfpu=vfp'
make[1]: *** [Kbuild:21: kernel/bounds.s] Error 1
```

I'll try native build later.

tmplt commented:
I'm trying to use these changes on a Raspberry Pi 3B to mirror the effect of a `dtparam=spi=on` (if I were on Raspbian), but I'm shooting a bit in the dark to be honest. On f0f88be1eabf2504736debb418d1f164b3228c8e with
```nix
# Unkown if needed, but read in some relevant thread that RPI overlays do not merge with mainline.
boot.kernelPackages = pkgs.linuxPackages_rpi3;

hardware.deviceTree = {
  enable = true;
  filter = "*-rpi-*.dtb";
  overlays = [{
    name = "spi";
    dtboFile = "${pkgs.device-tree_rpi.overlays}/spi0-hw-cs.dtbo";
}];
```
the image from `sd-image-aarch64.nix` builds and boots, but the build output did not contain any mention of overlay application, and I have no `/dev/spi*` files. The `dtb` file I presume is being used is `bcm2837-rpi-3-b.dtb` which contains `compatible = "raspberrypi,3-model-b\0brcm,bcm2837";` while the `dtbo` contains `compatible = "brcm,bcm2835";`. But after boot `dmesg` contains some `bcm2835` entries. Is there some minor revision mismatch only in the DT files? Is that why it didn't apply?

sorki commented:
@tmplt If there's no specific reason to use `linuxPackages_rpi3` you can use my SPI overlay with upstream kernel:

- https://github.com/hexamon-tech/nixos-lorawan-gateway/blob/master/modules/dts/spi.dts
- https://github.com/hexamon-tech/nixos-lorawan-gateway/blob/master/modules/kernel.nix#L11

If you switch compatible of the overlay to "raspberrypi" or "brcm,bcm2837" it would apply it.
I'm now a bit worried we need some fallback way to handle these unfortunate rpi foundation kernels and overlays.

tmplt commented:
@sorki I'm trying to build with that overlay using `nixpkgs.crossSystem.system = "aarch64-linux"`, but it seems `dtbs-with-symbols-aarch64-*` fails to build with `gcc: error: unrecognized command line option '-mlittle-endian'`. Do you use armv7l-linux in the configuration you linked? Does this derivation only build for a set of supported archs?

I also have a aarch64-linux cross-compilation in progress via binfmt, but it's taking ages to build on my local system. I don't know if it works yet, but it did not stumble upon the error mentioned above.

sorki commented:
@tmplt It contains cross compilation parts and it works for me with mainline kernels (yes I'm mostly using `armv7l`, will test on `aarch64` as well). Do you use it with mainline kernel or `_rpi` one? I've seen similar errors with `_rpi` kernel.

tmplt commented:
@sorki Mainline `linuxPackages_4_19`. I just tried with `linuxPackages_latest` which passes the build, but it does not boot: uboot reports `ERROR: Did not find a cmdline Flattened Device Tree`. I do not know if this is caused by the 5.8 kernel or the DT overlay.

tmplt commented:
> I also have a aarch64-linux cross-compilation in progress via binfmt

It finished, and it works: `spidev` and `spi_bcm2835` are loaded and `/dev/spi0.{0,1}` exists. There is some differences between the  expessions used for the binfmt build vis-a-vis the `nixpkgs.crossSystem.system = "aarch64-linux"` one. I'll dig them up if you need them.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/enabling-spi-on-the-raspberry-pi-3/9122/4

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/cross-compiling-problem-with-devicetree-on-armv6/9254/1

flokli commented:
I had another excursion into building device tree overlays with NixOS.

I wanted to use `GPIO_ACTIVE_LOW` and include `#include <dt-bindings/gpio/gpio.h>` for this.

After some time of debugging and `dtc` complaining about syntax errors, I realized `#include` directives are resolved in the kernel and u-boot by the C preprocessor, and dtc doesn't know anything about it.

I tried adding support for this into `nixos/modules/hardware/device-tree.nix`, essentially just running `$CC -E -I$kernelSrc/scripts/dtc/include-prefixes` before running `dtc`, but gave up in finding the right set of cli arguments to pass to $CC by staring at other Makefiles.

Maybe someone in here feels a bit more brave and wants to add this. It'd certainly be very useful.

leifhelm commented:
I have a device tree overlay (hifiberry-dacplus.dtbo from `raspberrypifw`) that is not compatible with `fdtoverlay`, but however it is compatible with `dtmerge` from `libraspberrypi`.

`ftdoverlay` exits with `FDT_ERR_NOTFOUND`.

I created my own device tree file and copied it to the `/boot/nixos/` directory and it worked.

On the `nixos-20.09` channel it is possible to say `base = ./.;` and `name = ./my.dtb`. Is there an equivalent configuration in the new `deviceTree` module?

flokli commented:
@leifhelm see https://github.com/NixOS/nixpkgs/pull/107637#issuecomment-752788529 - the raspberry-pi provided device tree overlays don't compile with the kernel-provided `dtc`, and the NixOS module doesn't have support to use other device tree compilers now.

As written in that comment, there's https://github.com/kwohlfahrt/nixpkgs/commit/7ba456e500addef74ca30f7bcefecbef1f94ae96, which could probably be upstreamed to nixpkgs and provide an alternative mechanism of compiling device tree overlays.

sorki commented:
> I tried adding support for this into `nixos/modules/hardware/device-tree.nix`, essentially just running `$CC -E -I$kernelSrc/scripts/dtc/include-prefixes` before running `dtc`, but gave up in finding the right set of cli arguments to pass to $CC by staring at other Makefiles.
> 
> Maybe someone in here feels a bit more brave and wants to add this. It'd certainly be very useful.

That would be welcome addition and was discussed on IRC some time ago. I might try to incorporate it to the next PR related to this and `raspberrypi` support - I'm considering adding something like `hardware.deviceTree.raspberryPi` that would use `dtmerge`.

sorki commented:
> I have a device tree overlay (hifiberry-dacplus.dtbo from `raspberrypifw`) that is not compatible with `fdtoverlay`, but however it is compatible with `dtmerge` from `libraspberrypi`.
> 
> `ftdoverlay` exits with `FDT_ERR_NOTFOUND`.
> 
> I created my own device tree file and copied it to the `/boot/nixos/` directory and it worked.
> 
> On the `nixos-20.09` channel it is possible to say `base = ./.;` and `name = ./my.dtb`. Is there an equivalent configuration in the new `deviceTree` module?

I think you can craft `hardware.deviceTree.package` manually. Module won't populate it if you set `hardware.deviceTree.enable = false` but it will be picked up by activation:

```nix
        ${optionalString (config.hardware.deviceTree.package != null) ''
          ln -s ${config.hardware.deviceTree.package} $out/dtbs
        ''}
```

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/boot-config-txt-related-questions-for-raspbarry-pi/3650/3

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/fdt-err-notfound-when-trying-to-apply-device-tree-overlay-on-raspberry-pi-4/19650/2
